w3-aggregation.md

```diff
@@ -0,0 +1,380 @@
+
+    Storefront->>Authority: invoke `aggregate/get`
+```
+
+## Capabilities
+
+This section describes the capabilities that form the w3 aggregation protocol, along with the details relevant for invoking capabilities with a service provider.
+
+In this document, we will be looking at `spade.storage` as an implementer of the `aggregate/*` and `offer/*` protocol.
+
+### `aggregate/offer`
+
+A Storefront principal can invoke a capabilty to submit an aggregate ready for offers.
+
+```iplsch
+type AggregateOffer union {
+  Aggregate  "aggregate/offer"
+} representation inline {
+  discriminantKey "can"
+}
+
+type Offer struct {
+  with   StorefrontDID
+  offer  { CARLink: OfferDetails }
+}
+
+type struct OfferDetails {
+    size        Int
+    commP       Bytes
+    src         [URL]
+}
+
+type CARLink string
+```
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/offer",
+    "nb": {
+      "offer": {
+        "link": "bafy...aggregate",
+        "size": 110101,
+        "src": ["https://w3s.link/ipfs/bafy...aggregate"]
+      }
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+This capability is invoked to submit a request to a broker service when an aggregate is ready for a Filecoin deal. The `nb.offer.link` is the DAG-CBOR CID that refers to a "Ferry" aggregate. It encodes a dag-cbor block with an array of entries representing all the CAR files to include in the aggregate. Its format is:
+
+```json
+[
+  {
+    "link": "bafy...",
+    "size": 110101,
+    "commP": "commP...",
+    "src": ["https://.../bafy(...)"]
+  } 
+]
+```
+
+A `src` with URLs to fetch the aggregate offer is provided. In addition, the total `size` of every CAR within it is provided, both for convenience and consistency with the dag-cbor blocks.
+
+A receipt will be generated to acknowledge the received request. This receipt MUST contain an [effect](https://github.com/ucan-wg/invocation/#7-effect) with a subsequent task (`.fx.join` field) that is run when submitted aggregate is processed and either succeeds (implying that aggregate was accepted and deals will be arranged) or fail (with `error` describing a problem with an aggregate)
+
+```json
+{
+  "ran": "bafy...invocation",
+  "out": {
+    "ok": {
+      "status": "queued"
+    }
+  },
+  "fx": {
+    "join": { "/": "bafy...dequeue" }
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+### `aggregate/get`
+
+A Storefront principal can invoke a capability to get state of a previously accepted aggregate.
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/get",
+    "nb": {
+      "link": "bafy...aggregate",
```

Comment:
We no longer have CID for the aggregate since we started inlining, and I'm not sure if deriving CID from the `nb.offer` to ask it's status is still making sense. Yet I do like deterministic identifier as opposed to made made up ones.

Reply:
This `bafy...aggregate` in here would be the CID of the CBOR block where we inline offer details. Maybe the issue is using the term aggregate, should we set `bafy.....offer` instead?

Reply:
ok wait, by your suggestion above now looks like we are not aligned in this. So, from my understanding of other meeting we would like to have all CARs to be part of the offer encoded in a single CBOR block, instead of within a CAR file with a block per entry.

With the above in mind, I would expect us to have a list with metadata of CARs (commP, cid, size), which would be encoded with DAG CBOR. Our service has an endpoint where this information can be fetched by the CID of the DAG CBOR.

In protocol: 

```
"nb": {
      "offer": {
        "link": "bafy...aggregate",
        "size": 110101,
        "src": ["https://w3s.link/ipfs/bafy...aggregate"]
      }
    }
```

`bafy...aggregate` would be the CID of the block which has information of all the CARs that are part of the offer. (perhaps we should call it `bafy...offer` to avoid confusion)


From [comment](https://github.com/web3-storage/specs/pull/51#discussion_r1163667443) it looks like @Gozala would be expected different approach where we just inline directly in `nb` the offer content (map with CID of each CAR to be included). The problem with that is that we do not have an identifier for the offer. We can derive it by calculating a CID for `nb`, but in my opinion that adds extra complexity. Moreover, offer content might be 10.000s of CAR files, which I would like to avoid bloating invocations size that go into the stream.

Reply:
Adding what I added in async convo:

in short, question here is how to represent offer sent from w3filecoin pipeline to spade-proxy in the UCAN protocol.
Offer contains the information for an aggregate of CAR files to be created. i.e a list of metadata of each CAR to be included (cid, commP, size, sourceUrl)

There are two alternatives
Encode offer with DAG CBOR, and provide in the UCAN capability invocation the cid for the offer, together with offer total size and sourceUrl
Inline in UCAN capability invocation metadata of all CAR files to be included

Expanding a bit, option 1 would look like:
"nb": {
  "offer": {
    "link": "bafy...offer",
    "size": 110101,
    "src": ["https://filecoin.web3.storage/bafy...offer"]
    }
}

While option 2 would look like

"nb": {
  "offer": {
      "bafy...car0": {
	       "size": 110101,
	        "commP": { "/": { "bytes": "7aEDQ...ZQQ" } },
	        "src": ["https://w3s.link/ipfs/bag...car0"]
      },
      "bafy...car1": {
	       "size": 110101,
	        "commP": { "/": { "bytes": "7aEDQ...ZQQ" } },
	        "src": ["https://w3s.link/ipfs/bag...car1"]
      }
      ...
  }
}

In option 1, we would get block that would have format of option 2.

With option 2 we can derive a CID as an identifier, by getting a CID for the nb object. While in option 1 the identifier CID is explicit